pkg/kube/krt/nestedjoin.go

```diff
@@ -0,0 +1,340 @@
+	for _, ts := range res {
+		m := j.merge(ts)
+		if m != nil {
+			l = append(l, *m)
+		}
+	}
+
+	return l
+}
+
+func (j *nestedjoin[T]) List() []T {
+	if j.merge != nil {
+		j.mergeList()
+	}
+
+	return j.quickList()
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *nestedjoin[T]) index(extract func(o T) []string) kclient.RawIndexer {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	ji := &dynamicJoinIndexer{indexers: make(map[collectionUID]kclient.RawIndexer)}
+	for _, c := range j.collections.List() {
+		ic := c.(internalCollection[T])
+		ji.indexers[ic.uid()] = ic.index(extract)
+	}
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		ji.Lock()
+		defer ji.Unlock()
+		switch e.eventType {
+		case collectionMembershipEventAdd:
+			ji.indexers[e.collectionValue.uid()] = e.collectionValue.index(extract)
+		case collectionMembershipEventDelete:
+			delete(ji.indexers, e.collectionValue.uid())
+		}
+	})
+	return ji
+}
+
+func (j *nestedjoin[T]) Synced() Syncer {
+	return channelSyncer{
+		name:   j.collectionName,
+		synced: j.synced,
+	}
+}
+
+func (j *nestedjoin[T]) HasSynced() bool {
+	return j.syncer.HasSynced()
+}
+
+func (j *nestedjoin[T]) WaitUntilSynced(stop <-chan struct{}) bool {
+	return j.syncer.WaitUntilSynced(stop)
+}
+
+func (j *nestedjoin[T]) Register(f func(o Event[T])) HandlerRegistration {
+	return registerHandlerAsBatched(j, f)
+}
+
+func (j *nestedjoin[T]) RegisterBatch(f func(o []Event[T]), runExistingState bool) HandlerRegistration {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	syncers := make(map[collectionUID]Syncer)
+	removes := map[collectionUID]func(){}
+	for _, c := range j.collections.List() {
+		reg := c.RegisterBatch(f, runExistingState)
+		ic := c.(internalCollection[T])
+		removes[ic.uid()] = reg.UnregisterHandler
+		syncers[ic.uid()] = reg
+	}
+	djhr := &dynamicJoinHandlerRegistration{
+		syncers: syncers,
+		removes: removes,
+	}
+
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		djhr.Lock()
+		defer djhr.Unlock()
+		switch e.eventType {
+		case collectionMembershipEventAdd:
+			reg := e.collectionValue.RegisterBatch(f, runExistingState)
+			djhr.removes[e.collectionValue.uid()] = reg.UnregisterHandler
+			djhr.syncers[e.collectionValue.uid()] = reg
+		case collectionMembershipEventDelete:
+			// Unregister the handler for this collection
+			remover := djhr.removes[e.collectionValue.uid()]
+			if remover == nil {
+				log.Warnf("Collection %v not found in %v", e.collectionValue.uid(), j.name())
+				return
+			}
+			remover()
+			delete(djhr.removes, e.collectionValue.uid())
+			delete(djhr.syncers, e.collectionValue.uid())
+
+			// Now send a final set of remove events for each object in the collection
+			// do this in a goroutine so we don't block running expensive handlers
+			go func() {
```

Comment:
I looked at this for a while and I may just have my nesting crossed (probably, tbh) so I'll just ask it: 

If we're doing "update=delete+add" type logic is it safe for a delete to be a long-tail process?  Basically, can we have:

1. delete processed
2. delete go routine starts sending some delete events
3. add processed
4. delete go routine still sending delete events for some amount of time

and wind up with a faulty update?

Reply:
This is a great point...I've been thinking that we didn't want to hold up processing for a slow event, but you're right that correctness is at stake. I'll move this out of the goroutine